### DISCUSSION TITLE
Remove particular pod instead of last broker (not on statefulset)

### DISCUSSION THREAD

**Discussion by kshitijmjoshi** [strimzi/strimzi-kafka-operator#8227] [Q&A]
This question is related to https://github.com/strimzi/strimzi-kafka-operator/issues/3527

We are on Strimzi latest (0.33.2) with Kafka version 3.4.0.

We have brokers 0 & 1 on gp2 volumes which we'd like to upgrade to gp3 (change the storage class) without downtime.

Since storageClass of PVs cannot be changed once they are created, I was planning to

1. move all the topics to broker 3,4 (already backed by gp3 volumes).
2. delete pods 0 and 1 and their PVCs.
3. recreate brokers 0 & 1 with new PVCs.

Since Strimzi no longer uses statefulSet, can a particular pod be taken down now instead of the last pod?

If not, is there a recommended method to change the storage class of particular pods? Thanks!

**Comment by scholzj** [comment]
If you use JBOD storage ... why don't you add new disk, move the data there and remove the old disk?

**Reply by kshitijmjoshi** [reply]
Unfortunately, we don't use JBOD storage.

**Reply by scholzj** [reply]
Well, with what we have in Strimzi today, this would require a little bit of hacking. You can scale to 4 nodes and move the data easily. But then you will need to hack the deletion of the brokers 0 and 1 and the deletion / recreation of the storage. This might work:

1) Scale to 4 brokers and move the topcis to brokers 2 and 3 as you planned
2) Stop the cluster operator (e.g. scale it to 0)
3) Delete the Pods and PVCs for brokers 0 and 1
4) Create manually the new PVCs for the brokers 0 and 1 with the new storage class
5) Restart the operator again and let it recreate the Pods for brokers 0 and 1
6) Move the topcis back to brokers 0 and 1
7) Scale back to 2 replicas

Additionally, unless you use default storage class, you will need to:
8) Edit the Kafka CR and change the storage class in the storage configuration to the new storage class
9) Edit the StrimziPodset for the Kafka brokers and edit the `strimzi.io/stroage` annotation and change the storage class there as well

----

Note: I never tried this. It should work, but try it in some test environment first.

**Reply by kshitijmjoshi** [reply]
Thanks, I'll try it out and post updates here.

**Reply by kshitijmjoshi** [reply]
This worked perfectly fine! No issues. Thanks for the directions.
Is there a recommendation for the zookeeper pods as well?

**Reply by scholzj** [reply]
How many ZooKeeper nodes you have? You can try something similar I guess -> 
* Make sure you have 3 or 5 ZooKeeper nodes
* Scale down the operator and delete the nodes one by one with the same process as for Kafka

ZooKeeper nodes are all active. So you cannot reassign the data. But you need majority of them to be online, so if you take it one by one, it should not cause availability issue. Just make sure you leave enough time between the individual nodes for them to resync the data.

(Again, not tested by myself recently, so you would need to test this)

**Reply by kshitijmjoshi** [reply]
Thanks! We have 3 nodes. I'll try it and post an update here.

**Reply by kshitijmjoshi** [reply]
Worked fine with no downtime. Thanks for the guidance!

**Reply by scholzj** [reply]
Great, glad it worked. Thanks for confirming.
